075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+
+* Increasing the content size of the `Kafka` custom resource which could be a limit for future additions (even not related to rebalancing).
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...  
+    autoRebalance:
+      # using the same rebalancing configuration when adding or removing brokers ... a lot of duplication
+      - mode: add-brokers
+        goals:
+          - CpuCapacityGoal
+          - NetworkInboundCapacityGoal
+          - DiskCapacityGoal
+          - RackAwareGoal
+          - MinTopicLeadersPerBrokerGoal
+          - NetworkOutboundCapacityGoal
+          - ReplicaCapacityGoal
+        skipHardGoalCheck: true
+        # ... other rebalancing related configuration
+      - mode: remove-brokers
+        goals:
+          - CpuCapacityGoal
+          - NetworkInboundCapacityGoal
+          - DiskCapacityGoal
+          - RackAwareGoal
+          - MinTopicLeadersPerBrokerGoal
+          - NetworkOutboundCapacityGoal
+          - ReplicaCapacityGoal
+        skipHardGoalCheck: true
+```
+
+### Alternative 3
+
+This alternative is about not leveraging a `KafkaRebalance` custom resource, automatically created by the Cluster Operator, but having a more tight integration between the Cruise Control rebalancing FSM (Finite State Machine) and the `KafkaAssemblyOperator`.
+It means that we should factor out the Cruise Control rebalancing FSM (and all the corresponding interaction with the Cruise Control API) from the current `KafkaRebalanceAssemblyOperator`.
+This way, the Cruise Control rebalancing FSM could be used directly by both the `KafkaRebalanceAssemblyOperator` for the user initiated rebalancing as today and the `KafkaAssemblyOperator` for the auto-rebalancing feature (a reference example could be the KRaft migration FSM).
+
+In the first scenario, the `KafkaRebalanceAssemblyOperator` should be able to interact directly with the Cruise Control rebalancing FSM by getting any rebalancing parameter from the `KafkaRebalance` custom resource created by the user.
+In the second scenario, the `KafkaAssemblyOperator` should be able to interact directly with the Cruise Control rebalancing FSM by getting any rebalancing parameter from a settings resource or from the Cruise Control defaults, without any need for an auto-created `KafkaRebalance` custom resource which is the main goal of this alternative.
+
+Pros:
+
+* No automatic `KafkaRebalance` custom resource creation.
+* No user confusion between auto-created and manually-created `KafkaRebalance` custom resources.
+
+Cons:
+
+* More complex and more work on factoring out the Cruise Control rebalancing FSM from the `KafkaRebalanceAssemblyOperator`.
+* Not leveraging the already ready-to-use `KafkaRebalanceAssemblyOperator` through a `KafkaRebalance` custom resource.
+* `KafkaAssemblyOperator` strongly tight to the Cruise Control rebalancing FSM with additional complexity for testing.
```

Comment:
I agree that for this feature it would be simplest to reuse the logic of the `KafkaRebalanceAssemblyOperator` through the `KafkaRebalance` resource **BUT** in the long run it would be wiser to factor out the CC rebalancing FSM first, decoupling it from the `KafkaRebalanceAssemblyOperator` for the following reasons:

(1) It would help us consolidate the CC client code used by other Strimzi components like the Topic Operator (from what I understood we were planning on moving toward this direction anyway after the replication factor work)

(2) It would simplify the existing `KafkaRebalanceAssemblyOperator` which is already pretty complicated and difficult to debug/extend

(3) It would simplify the addition of new features that leverage the CC API - especially the CC Self-Healing support which is the next logical feature after we nail the auto-balancing feature.

If we don't clean this now it'll be harder to in the future because more features will be dependent on it. It'll not be an easy job to refactor but if people agree to go in this direction I will help with it and suffer with you or whoever decides to implement the autobalancing.

Reply:
The alternative 3 was mostly rejected because the last point, not because of factoring out the rebalance FSM.
If we want, it's something we can still do but unrelated to this proposal.
I mean the bigger problem is having the `KafkaAssemblyOperator` using directly the FSM which would add more complexity for testing. This is why we would like to leverage the rebalancing we have today through the `KafkaRebalanceAssemblyOperator`.
But again, doing this way, doesn't mean that we can't factor out the rebalancing FSM. We can do that but it will be still used in the `KafkaRebalanceAssemblyOperator` only and not within the `KafkaAssemblyOperator` for auto-rebalancing purposes.

Reply:
The CC client sharing is still in my backlog. I will come to it as soon as pending TO refactoring tasks are completed.

Reply:
But the CC client sharing is totally different from factoring out the rebalancing FSM. They are two different and unrelated things.

Reply:
Yes, I agree. Just wanted to confirm it's still in the plan. Factoring out the rebalancing FSM would require a dedicated proposal, like the CC client sharing.

Reply:
> I mean the bigger problem is having the KafkaAssemblyOperator using directly the FSM which would add more complexity for testing. This is why we would like to leverage the rebalancing we have today through the KafkaRebalanceAssemblyOperator.

You both are right, factoring out the rebalance FSM is a big piece of work that merits its own PR.

From what I understand, this obstacle is what prevents us from easily using the rebalance FSM directly for the auto-rebalancing work.

A couple of things I have been mulling over:

(1) Is the creation, manipulation, and management of custom resource content supposed to be in the domain of the operator or just the end users? 

(2) For adding support for other Cruise Control features like self-healing or brokers demotion, the operator would benefit from being able to hit rebalance FSM directly. If we had these features hit the rebalance FSM directly, would we want to update the autoscale rebalance feature to hit the rebalance FSM directly as well at a later date?

(3) Although the CC client sharing work is a separate task, are there any ideas from the Topic Operator replication factor work that we can reuse here? I would imagine that the CC client code used by the TO  would have solved challenges similar to the ones we will face with autorebalancing and other CC features we may hope to support in the future

Reply:
> From what I understand, this obstacle is what prevents us from easily using the rebalance FSM directly for the auto-rebalancing work.

No. The thing is that using rebalance FSM directly in the `KafkaAssemblyOperator` means tightening it to rebalance stuff and making less testable. 

Regarding (1) ... not sure I got what you mean.
Regarding (2) ... see my first answer. Not sure how self-healing would been added but I don't think it would be within the `KafkaAssemblyOperator`
Regarding (3) ... I think CC client sharing doesn't have any impact on the auto-rebalancing. We are currently using the one in the cluster operator. One day we'll use the common one. In the end it's just the client for calling the CC REST API.

Reply:
> Regarding (1) ... not sure I got what you mean.

I mean to ask: Were custom resources in general were designed to be created and configured by K8s Operators and not just by users? Furthermore, I wonder if Strimzi custom resources were in this way too, and if not, are there are any obvious consequences of doing so. I need to think more on this

> Regarding (2) ... see my first answer. Not sure how self-healing would been added but I don't think it would be within the KafkaAssemblyOperator

I interpreted Alternative 3 more as: Hitting the CC API directly instead of using the `KafkaRebalance` resource. Despite the immediate obstacles of doing so I was thinking that it may be easier route to follow for the long run. Let me think more on this, I haven't looked that deep into the KAO or KRBAO code yet

Reply:
> Hitting the CC API directly instead of using the KafkaRebalance resource

And your understanding is right but it's unrelated to the CC API client refactoring.
As I mentioned, alternatives 3 brings KAO to be strictly tight to CC which is something we want to avoid because it's focused on handling `Kafka` custom resources not calling other external stuff. It's the role of the KRAO by handling `KafkaRebalance`. Also testing KAO would be more difficult.